src/rdb/rdb_raft.c

```diff
@@ -1237,10 +1252,36 @@ rdb_raft_cb_log_offer(raft_server_t *raft, void *arg, raft_entry_t *entries,
 	if (!db->d_raft_loaded)
 		return 0;
 
-	for (i = 0; i < *n_entries; ++i) {
-		rc = rdb_raft_log_offer_single(db, &entries[i], index + i);
-		if (rc != 0)
+	/*
+	 * Conservatively employ one VOS TX for each entry for now, so that if
+	 * an entry encounters an error, we still end up making some progress
+	 * by not rolling back prior entries in the batch. Once VOS supports
+	 * batching TXs, we can optimize this process further.
+	 */
+	for (i = 0; i < *n_entries; i++) {
+		rdb_vos_tx_t vtx;
+
+		rc = rdb_raft_entry_count_vops(db, &entries[i]);
+		if (rc < 0) {
+			DL_ERROR(rc, DF_DB ": failed to count VOS operations", DP_DB(db));
+			break;
+		}
+
+		rc = rdb_vos_tx_begin(db, rc, &vtx);
+		if (rc != 0) {
+			DL_ERROR(rc, DF_DB ": failed to begin VOS TX for entry %ld", DP_DB(db),
+				 index);
+			break;
+		}
+
+		rc = rdb_raft_log_offer_single(db, &entries[i], index + i, vtx);
+
+		rc = rdb_vos_tx_end(db, vtx, rc);
+		if (rc != 0) {
+			DL_ERROR(rc, DF_DB ": failed to end VOS TX for entry %ld", DP_DB(db),
+				 index);
```

Comment:
Do you know what's the typical transaction size after this change?

FYI, there is a potential bug in PMDK which could be triggered when the transaction redo log is large. (see DAOS-13081 and https://github.com/pmem/pmdk/issues/5583), but seems we never hit the issue so far, I guess it because the redo log size (not including snapshot which is in undo log) is usually very small for PMDK transaction.